wave_calculator.gd

```diff
@@ -6,19 +6,50 @@ var rng: RandomNumberGenerator = RandomNumberGenerator.new()
 var _base_spawn_time: float
 var _enemies: Dictionary  # Was not able to type the dict either
 
+const MAX_WAVE_BATCH_SCORE = 10
+
 
 func _init(base_spawn_time: float, enemies: Dictionary) -> void:
 	_base_spawn_time = base_spawn_time
 	_enemies = enemies
 
 
 # Array[Dictionary] doesn't work :sad: https://forum.godotengine.org/t/array-with-dictionaries-doesnt-get-recognized-as-array-dictionary/58280
-func get_enemy_spawns(wave_number: int) -> Array:
+func get_enemy_spawns(wave_number: int):
 	var difficulty = _get_difficulty(wave_number)
+	var total_batches = max(1, ceil(float(difficulty) / MAX_WAVE_BATCH_SCORE))
+
+	var result = []
+	var total_difficulty = 0
+
+	for _index in range(total_batches):
+		var enemy_difficulty_pool = _enemies.keys().filter(
+			_filter_lte(_get_max_enemy_score(difficulty))
+		)
+		var max_enemy_difficulty = enemy_difficulty_pool.max()
+
+		var batch = []
+		var batch_score = 0
+		var target_batch_score = min(difficulty - total_difficulty, MAX_WAVE_BATCH_SCORE)
+
+		while batch_score < target_batch_score and total_difficulty < difficulty:
+			var remaining_batch_score = target_batch_score - batch_score
+			if remaining_batch_score < max_enemy_difficulty:
+				enemy_difficulty_pool = enemy_difficulty_pool.filter(
+					_filter_lte(remaining_batch_score)
+				)
+				max_enemy_difficulty = enemy_difficulty_pool.max()
 
-	var enemy_scores = _get_enemy_scores(difficulty)
+			var enemy_difficulty = enemy_difficulty_pool.pick_random()
 
-	return enemy_scores.map(_add_spawn_timer)
+			batch_score += enemy_difficulty
+			total_difficulty += enemy_difficulty
+
+			batch.append(enemy_difficulty)
+
+		result.append(batch)
```

Comment:
I think this entire thing can get A LOT more complicated if you extract a function `get_random_enemy_max_score(int)` that fetches a random enemy with that max. Which can let you not recompute the `enemy_difficulty_pool` and all that.

Reply:
moved things to a separate `_batch` function which simplified things. we prune the max in an enemy score list arg that get's duplicated so we can mutate it as we pick random enemy scores from the pool.